File: 2025/200. Number of Islands/Solution.java
class Solution {
    
    int dx[] = {0, 1, 0, -1};
    int dy[] = {1, 0, -1, 0};

    public int numIslands(char[][] grid) {
        int n = grid.length, m = grid[0].length;
        int count = 0;
        for(int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == '1') {
                    count++;
                    traverse(new int[]{i, j}, grid);
                }
            }
        }
        return count;  
    }

    private void traverse(int[] point, char[][] grid) {
        int x = point[0], y = point[1];
        if (x < 0 || y < 0 || x >= grid.length || y >= grid[0].length || grid[x][y] != '1')
            return;
        grid[x][y] = '0';
        for (int i = 0; i < 4; i++) {
            traverse(new int[]{x + dx[i], y + dy[i]}, grid);
        }
    }
}

File: 2025/200. Number of Islands/solution.py
class UnionFind:
    def __init__(self, size: int):
        self.size = size
        self.parent = [i for i in range(size)]

    def find(self, x: int) -> int:
        if x != self.parent[x]:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x: int, y: int):
        root_x = self.find(x)
        root_y = self.find(y)
        if root_x != root_y:
            self.parent[root_x] = root_y
            self.size -= 1

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m, n = len(grid), len(grid[0])
        unionfind = UnionFind(m * n)
        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1":
                    for (x,y) in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
                        ii, jj = i + x, j + y
                        if ii >= 0 and jj >= 0 and ii < m and jj < n and grid[ii][jj] == "1":
                            unionfind.union(i * n + j, ii * n + jj)
                else:
                    unionfind.size -= 1
        return unionfind.size